viamonvif/mdns_server.go

```diff
@@ -82,6 +102,89 @@ func (server *mdnsServer) Remove(hostname string) {
 	delete(server.mappedDevices, hostname)
 }
 
+type cachedEntry struct {
+	DNSName string `json:"dnsname"`
+	IP      string `json:"ip"`
+}
+
+func (server *mdnsServer) readCacheFile() []cachedEntry {
+	// If the `cacheFilepath` was given, read the contents and parse it as json. On any error we log
+	// and return nil, which is interpreted as an empty slice.
+	if server.cacheFilepath == "" {
+		return nil
+	}
+
+	file, err := os.ReadFile(server.cacheFilepath)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+
+		server.logger.Warn("Error reading dns cache file. File:", server.cacheFilepath, "Err:", err)
+		return nil
+	}
+
+	var results []cachedEntry
+	err = json.Unmarshal(file, &results)
+	if err != nil {
+		server.logger.Warn("Error unmarshalling JSON. File:", server.cacheFilepath, "Err:", err)
+		return nil
+	}
+
+	return results
+}
+
+func (server *mdnsServer) apply(newEntries []cachedEntry) {
+	// For each entry (presumably from the `cachedFilepath`), call `mdnsServer.Add`.
+	for _, entry := range newEntries {
+		ip := net.ParseIP(entry.IP)
+		if ip == nil {
+			server.logger.Warn("Unknown IP from cache file. DNSName:", entry.DNSName, "IP:", entry.IP)
```

Comment:
[q] worth running `DiscoverResources` in the background when the resource is configured as well? so if an ip address gets changed there will be less downtime for the user.

Reply:
Though I guess if we did that we wouldn't really need to have the cache file.

Reply:
I think a cache file largely eliminates the need, yes. But I do want to explore what you're describing with https://viam.atlassian.net/browse/RSDK-10102